File: java/src/main/java/io/binac/leetcode/LongestCommonPrefix.java
package io.binac.leetcode;

/**
 * <p>Write a function to find the longest common prefix string amongst an array of strings.</p>
 * <p>If there is no common prefix, return an empty string <code>""</code>.</p>
 * <p>&nbsp;</p>
 * <p><strong>Example 1:</strong></p>
 * <pre><strong>Input:</strong> strs = ["flower","flow","flight"]
 * <strong>Output:</strong> "fl"
 * </pre>
 * <p><strong>Example 2:</strong></p>
 * <pre><strong>Input:</strong> strs = ["dog","racecar","car"]
 * <strong>Output:</strong> ""
 * <strong>Explanation:</strong> There is no common prefix among the input strings.
 * </pre>
 * <p>&nbsp;</p>
 * <p><strong>Constraints:</strong></p>
 * <ul>
 * <li><code>0 &lt;= strs.length &lt;= 200</code></li>
 * <li><code>0 &lt;= strs[i].length &lt;= 200</code></li>
 * <li><code>strs[i]</code> consists of only lower-case English letters.</li>
 * </ul>
 */
public class LongestCommonPrefix {
    public static class Solution1 {
        public String longestCommonPrefix(String[] strs) {
            if (strs.length == 0)
                return "";

            char ch;
            for (int j = 0, n = strs[0].length(), i, m = strs.length; j < n; ++j) {
                ch = strs[0].charAt(j);
                for (i = 1; i < m; ++i) {
                    if (j >= strs[i].length() || ch != strs[i].charAt(j))
                        return strs[0].substring(0, j);
                }
            }
            return strs[0];
        }
    }
}

File: python3/leetcodepy/longest_common_prefix.py
"""
Write a function to find the longest common prefix string amongst an array of strings.

If there is no common prefix, return an empty string "".



Example 1:

Input: strs = ["flower","flow","flight"]
Output: "fl"

Example 2:

Input: strs = ["dog","racecar","car"]
Output: ""
Explanation: There is no common prefix among the input strings.


Constraints:

0 <= strs.length <= 200
0 <= strs[i].length <= 200
strs[i] consists of only lower-case English letters.
"""
from typing import List


class Solution1:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if not strs:
            return ''
        for j in range(len(strs[0])):
            ch = strs[0][j]
            for i in range(1, len(strs)):
                if j >= len(strs[i]) or strs[i][j] != ch:
                    return strs[0][:j]
        return strs[0]
